In [1]:
!pip install langchain groq
!pip install langchain-groq groq
!pip install openai


     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 50.4/50.4 kB 1.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 997.8/997.8 kB 20.2 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 103.5/103.5 kB 6.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 76.4/76.4 kB 3.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 77.9/77.9 kB 4.3 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 394.9/394.9 kB 16.9 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 150.5/150.5 kB 9.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 141.9/141.9 kB 9.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 58.3/58.3 kB 2.6 MB/s eta 0:00:00
  Attempting uninstall: tenacity
    Found existing installation: tenacity 9.0.0
    Uninstalling tenacity-9.0.0:
      Successfully uninstalled tenacity-9.0.0
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 362.9/362.9 kB 11.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━

Q1 Implement a Chain-of-Thought prompting system to solve word problems. Your system should break down the problem-solving process into steps and show its reasoning. Task: ● Create a function that takes a word problem as input. ● Use the OpenAI API or any other LLM API to generate a step-by-step solution. ● The output should show each step of the reasoning process. ● Test your function with at least two word problems of varying complexity. Example word problem: "A baker has 150 eggs. If each cake requires 3 eggs and each cookie requires 2 eggs, how many cakes and cookies can the baker make if they want to use all the eggs and make an equal number of each?"

In [2]:
from langchain_groq import ChatGroq
from langchain.chains import LLMChain, SimpleSequentialChain
from langchain_core.messages import AIMessage, HumanMessage, SystemMessage
from langchain.prompts import Prompt
from langchain import PromptTemplate, LLMChain
import os

# Set your Groq API key
os.environ["GROQ_API_KEY"] = "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"

# Initialize the Groq LLM
llm = ChatGroq(model_name="mixtral-8x7b-32768")


In [3]:
example = """
Question:" Sean buys 25 stickers on Monday and 17 on Tuesday. On Wednesday, he gives 5/6
 th of his stickers to James. How many does he have left?"

"""
human_message = HumanMessage(content=example)
system_message = SystemMessage(content="You are helpful assistant.")


new_prompt = (system_message + human_message)

print (f"Prompt: {new_prompt}")

chain = LLMChain(llm=llm, prompt=new_prompt)
reponse = chain.run({})
print(f"Response: {reponse}")


Prompt: input_variables=[] messages=[SystemMessage(content='You are helpful assistant.'), HumanMessage(content='\nQuestion:" Sean buys 25 stickers on Monday and 17 on Tuesday. On Wednesday, he gives 5/6\n th of his stickers to James. How many does he have left?"\n\n')]


/usr/local/lib/python3.10/dist-packages/langchain_core/_api/deprecation.py:151: LangChainDeprecationWarning: The class `LLMChain` was deprecated in LangChain 0.1.17 and will be removed in 1.0. Use RunnableSequence, e.g., `prompt | llm` instead.
  warn_deprecated(
/usr/local/lib/python3.10/dist-packages/langchain_core/_api/deprecation.py:151: LangChainDeprecationWarning: The method `Chain.run` was deprecated in langchain 0.1.0 and will be removed in 1.0. Use invoke instead.
  warn_deprecated(


Response: First, let's find out how many stickers Sean had in total on Monday and Tuesday. 

Sean bought 25 stickers on Monday and 17 on Tuesday, so he had 25 + 17 = 42 stickers in total.

On Wednesday, Sean gave 5/6 of his stickers to James. To find out how many stickers he gave to James, we need to divide the total number of stickers by 6 and then multiply the result by 5 (because 5/6 is the same as multiplying by 5 and then dividing by 6):

Stickers given to James = (Total stickers / 6) * 5 = (42 / 6) * 5 = 7 * 5 = 35/6 ≈ 5.83

Since Sean can't give a fraction of a sticker, we'll round this number to the nearest whole number, which is 6.

Therefore, Sean has 42 - 6 = 36 stickers left after giving some to James.


In [4]:
example="""
Question:"There are 235 books in a library. On Monday, 123 books are taken out. On Tuesday, 56 books are brought back. How many books are there now?"
"""

human_message = HumanMessage(content=example)
system_message = SystemMessage(content="You are helpful assistant.")


new_prompt = (system_message + human_message)

print (f"Prompt: {new_prompt}")

chain = LLMChain(llm=llm, prompt=new_prompt)
reponse = chain.run({})
print(f"Response: {reponse}")

Prompt: input_variables=[] messages=[SystemMessage(content='You are helpful assistant.'), HumanMessage(content='\nQuestion:"There are 235 books in a library. On Monday, 123 books are taken out. On Tuesday, 56 books are brought back. How many books are there now?"\n')]
Response: To find out how many books are in the library now, we first need to find out how many books are in the library after Monday. This would be the total number of books before Monday minus the number of books taken out on Monday.

Books left after Monday = Total books - Books taken out
= 235 books - 123 books
= 112 books

Then, on Tuesday, 56 books were brought back. So, we add this number to the number of books left after Monday to find out how many books are in the library now.

Books in the library now = Books left after Monday + Books brought back
= 112 books + 56 books
= 168 books

So, there are 168 books in the library now.


Q 2)Develop a prompt-chaining system that generates a short story based on user input. The system should use multiple prompts in sequence, with each prompt building on the output of the previous one. Task: ● Create a series of at least 3 prompts that will be chained together. ● The first prompt should ask for a genre and main character. ● Subsequent prompts should build on previous outputs to develop the plot and conclusion. ● Implement the chaining logic to pass information between prompts. ● Allow the user to input the initial genre and main character. ● Output the final story generated through the chain

In [5]:
system_message = SystemMessage(content="You are a creative assistant for generating short stories.")

# Prompt 1: Get genre and main character
initial_prompt = HumanMessage(content="Provide a genre and a main character for a short story.")

# Create the chain and get the genre and character
chain1 = LLMChain(llm=llm, prompt=system_message + initial_prompt)
response1 = chain1.run({})

try:
    genre, character = response1.split(', ')
except ValueError as e:
    print(f"Error parsing response: {e}")
    genre, character = None, None


response1 = "Action, Avengers:End Game "
print(f"Genre: {genre}, \nCharacter: {character}")


plot_prompt = HumanMessage(content=f"Based on the genre '{genre}' and the main character '{character}', write a brief plot outline for a short story.")


chain2 = LLMChain(llm=llm, prompt=system_message + plot_prompt)
plot = chain2.run({})


conclusion_prompt = HumanMessage(content=f"Based on the plot outline provided, write the conclusion for the short story.")


chain3 = LLMChain(llm=llm, prompt=system_message + conclusion_prompt)
conclusion = chain3.run({})


final_story = f"Genre: {genre}\nMain Character: {character}\nPlot: {plot}\nConclusion: {conclusion}"
print(final_story)

Error parsing response: not enough values to unpack (expected 2, got 1)
Genre: None, 
Character: None
Genre: None
Main Character: None
Plot: Title: The Infinite Blank

In a world where stories and their characters are bought and sold as commodities, a unique, unclaimed character named None finds itself in the Infinite Blank - an endless, formless void that serves as the birthplace of all narratives. 

None, being neither good nor evil, male nor female, or belonging to any particular genre, is a mystery to the other characters and story-makers. This lack of definition gives None the power to choose its own destiny, and the ability to explore the Infinite Blank without limitations.

During its journey, None encounters various story elements, such as genre tropes, plot devices, and character archetypes, scattered throughout the void. As None interacts with these elements, it begins to understand the complexity and versatility of stories. 

As None collects these story pieces, it starts to

Q3)Implement a Tree-of-Thought prompting system to solve a complex problem. The system should explore multiple paths of reasoning and select the most promising one. Task: ● Create a function that takes a complex problem as input. ● Generate at least 3 different initial approaches to solving the problem. ● For each approach, develop 2-3 possible next steps. ● Implement a method to evaluate the promise of each path. ● Select the most promising path and generate a final solution. ● Test your system with the following problem: Problem: "Design a sustainable urban transportation system for a city of 1 million people. Consider factors such as environmental impact, cost-effectiveness, and accessibility for all residents

In [6]:
first_prompt = PromptTemplate(
    input_variables=["topic"],
    template="Write a short introduction about {topic}."
)

# Second prompt template
second_prompt = PromptTemplate(
    input_variables=["introduction"],
    template="Based on this introduction, suggest 3 key points to expand upon: {introduction}"
)

# Third prompt template
third_prompt = PromptTemplate(
    input_variables=["key_points"],
    template="For each of these key points, provide a brief explanation: {key_points}"
)

# Create chains
chain_one = LLMChain(llm=llm, prompt=first_prompt)
chain_two = LLMChain(llm=llm, prompt=second_prompt)
chain_three = LLMChain(llm=llm, prompt=third_prompt)

# Create the sequential chain
overall_chain = SimpleSequentialChain(
    chains=[chain_one, chain_two, chain_three],
    verbose=True
)

# Run the chain
topic = "future of ai in machine learninng"
result = overall_chain.run(topic)

print(result)



> Entering new SimpleSequentialChain chain...
The future of Artificial Intelligence (AI) in machine learning is promising and holds the potential to revolutionize various industries. AI algorithms are expected to become more sophisticated, enabling machines to learn from data and make decisions with greater accuracy and speed. Advances in deep learning and neural networks will allow for more complex problem-solving and pattern recognition. Moreover, the integration of AI with other technologies such as IoT, blockchain, and quantum computing will result in even more powerful and intelligent systems. However, there are also challenges and ethical considerations that need to be addressed, such as data privacy, bias, and job displacement. Overall, the future of AI in machine learning is bright, but it is crucial to approach it with caution and responsibility.
1. Enhanced AI algorithms for improved decision-making: As AI algorithms become more sophisticated, machines will be able to learn